### ISSUE TITLE
[3] Node selection latency and up/down indicators are incorrect

### ISSUE THREAD

**Issue by btswolf** [bitshares/bitshares-ui#376]
The node selection site does not work properly.
Nodes are often displayed as OFF even if they are obviously ON.
Latency values do not makes sense, should be the value of an ICMP, don't know what is measured here but latency of over 900 for most nodes cannot be correct.
[![image](https://user-images.githubusercontent.com/11035520/30251546-a159b276-9661-11e7-9397-9502fcee93ae.png)](https://user-images.githubusercontent.com/11035520/30251546-a159b276-9661-11e7-9397-9502fcee93ae.png)

**Comment by wmbutler** [comment]
I allocated a couple of hours to this because as I understand it, we are simply reporting the backend values that we receive. It may take some time to understand why this endpoint is reporting incorrect values. I do think this is an important bug to resolve. Once again, it speaks to confidence.

**Comment by svk31** [comment]
Latencies will definitely be different from simple pings, since they measure the time taken to negotiate a websocket connection with the witness_node API. The websocket may even be hosted on a different server than the domain name and just redirected using nginx for example. The fact that the domain is up also does not mean that the websocket is up, it could be down, or incorrectly configured. For performance reasons we do not poll all the websocket endpoints every time you load the GUI, only every 15 times.

What we could do is trigger a latency check on all API endpoints when you go to the Access page in the settings.

**Comment by btswolf** [comment]
I understand that latency may be different from simple pings but the GUI should not indicate that most nodes are down or have bad latency if that is obviously not the case. If latency up to 250ms for example is good/average that should be green not yellow.
A full poll on each GUI start may not be required but a background poll every 15min could help to find a better node if the current connection is yellow/red. And if a better connection has been found a prompt could ask the user to reconnect or to stay connected for the current session.

**Comment by wmbutler** [comment]
A factor of 4x off is pretty large. Definitely worth digging into this.

**Comment by svk31** [comment]
You're comparing apples to oranges so there's no reason to say 4x difference is large. And no, light client or web client does not matter, it's always measured from your local computer.

I do agree there seems to be an issue with nodes being marked as down when they're not, that can probably be solved by my earlier suggestion.

**Comment by wmbutler** [comment]
Are you saying that ICMP vs socket latency has that much variation? That's not been my experience so I'm surprised to hear you say that. The difference between 40ms and 160ms is pretty large. So large that it makes me think it's being calculated incorrectly. Possibly roundtrip x 4 for some reason. I just think it's worth investigating.

**Comment by wmbutler** [comment]
70ms vs 743 ms for dexnode.net

![screen shot 2017-09-11 at 2 04 43 pm](https://user-images.githubusercontent.com/1254810/30292029-3560093c-96fa-11e7-9003-c0d35f97b943.png)
![screen shot 2017-09-11 at 2 04 27 pm](https://user-images.githubusercontent.com/1254810/30292030-35620426-96fa-11e7-99c8-ee8033a5edbd.png)

**Comment by svk31** [comment]
Like I tried to explain above; you are measuring two completely different things, so it's normal that the latencies differ.

**Comment by wmbutler** [comment]
Can we change how this value is derived? Measuring the time to negotiate a WebSocket is not an accurate representation of how responsive a remote node is.

**Comment by svk31** [comment]
What makes you think it isn't? Measuring the ping of an ip that may or may not have anything to do with the websocket, like what btswolf is doing, is certainly not an accurate representation of the node.

**Comment by wmbutler** [comment]
Well, in my 15 years of building web applications, I've always considered response times of > 150ms as unacceptable. I think that 750ms more than likely represents a load time, possibly the time it takes to load static files from the the web site, but it seems unlikely that it's an accurate representation of socket latency. We run sockets with our web application and they are sub 50ms in almost all cases. I'm not trying to be argumentative here, but it would be good to understand exactly what is being measured here. @xeroc do you have any words of wisdom to contribute to this conversation? @calvinfroedge ?

**Comment by calvinfroedge** [comment]
I think there may be a balance that can be struck. For instance, perhaps time to do a web socket heartbeat will be less than time to negotiate initial connection, and more representative of responsiveness once the connection is open?

I do agree with @svk31 that ping is kind of useless if it's not what's actually being used to handle work.

**Comment by wmbutler** [comment]
Totally agree that latency is the proper term.

**Comment by wmbutler** [comment]
Pushing this to next Sprint and increasing the bounty.

**Comment by svk31** [comment]
I've identified a few bugs in the api server fallback flow and fixed those, and also noticed that when a fallback was attempted, the latency of the successful node would be set to the time it took to try all the fallbacks so I fixed that.

![image](https://user-images.githubusercontent.com/6890015/30691625-a3ceff88-9ec8-11e7-8035-dac30bf04e4a.png)

I also added "Active" status to the automatic selection so it's clear that it is being used, and removed the status for the testnet server since it always says down.

The actual latency values aren't important, perhaps we should just name them "low", "medium" and "high" so users don't get caught up in the actual values?
